### ISSUE TITLE
C Concurrency in Action 笔记

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11189]
C++ Concurrency in Action 笔记<br>
<br>
https://ift.tt/A5OuNji<br>
<br>
<br>
<br>
<div><div>
<p>记录一下我从《C++ Concurrency in Action, 2nd Edition》中学到的东西。</p>
<p>注：不怎么全，漏掉的可能是我没学会的，也可能是我本来就会的，还可能是我觉得用不上的。</p>
<span></span>
<h2>Chapter 1</h2>
<p>为什么要使用并发/为什么不要使用并发？</p>
<ul>
<li>使用并发的理由主要是划分业务和提升性能。提升性能很好理解，划分业务主要指的是另开线程用于处理不确定何时会发生的事件（响应用户请求，监听网络数据之类的）。</li>
<li>不使用并发的理由是“这样做不值得”，并发带来的复杂度胜过了带来的性能提升（甚至性能也不一定有提升）。</li>
</ul>
<h2>Chapter 2</h2>
<p><code>std::thread</code> 相关的坑？</p>
<ul>
<li>C++'s most vexing parse，也就是把变量定义当作函数声明。所以要尽量用大括号初始化。（相比起运行时的罕见 bug，编译错误真的是小问题。）</li>
<li>Dangling reference/pointer（悬挂引用/悬挂指针）。这是涉及到闭包时的经典问题，lambda 捕获引用时注意一下被捕获对象的生命周期。</li>
<li>在 <code>std::thread</code> 对象被销毁前记得调用 <code>.detach()</code> 或者 <code>.join()</code>，否则 <code>std::thread</code> 析构函数里会帮你调 <code>std::terminate()</code>，然后你所有 thread 就全没了。
<ul>
<li>此外也要注意代码中的 <code>.join()</code> 会不会因为异常而没被执行。</li>
</ul>
</li>
<li>如果要靠 <code>std::thread</code> 传函数参数的话，引用需要用 <code>std::ref</code>/<code>std::cref</code> 包一下。（和 <code>std::bind</code> 的情况一样，更推荐使用 lambda，清晰自然。）</li>
</ul>
<h2>Chapter 3</h2>
<p><code>std::lock_guard</code>, <code>std::scoped_lock</code> 和 <code>std::unique_lock</code> 的区别？</p>
<ul>
<li>
<code>std::scoped_lock</code> 可以完美替代 <code>std::lock_guard</code>，甚至支持 0 个或者多个锁，而 <code>std::lock_guard</code> 只支持一个。</li>
<li>
<code>std::unique_lock</code> 可以在构造时不上锁（通过第二个参数指定）以及在作用域内释放锁（支持 <code>.unlock()</code>），通常和 <code>std::condition_variable</code> 的 <code>.wait()</code> 一起服用。</li>
</ul>
<p>一个关于 stack API 设计的坑：</p>
<ul>
<li>以前我一直不明白为啥 <code>std::stack/queue</code> 之类的 STL 的容器的 <code>.pop()</code> 方法不带返回值，用起来不大方便，因为很多时候都是 <code>.top()</code> 和 <code>.pop()</code> 连用的。其中的原因在于，在返回值给用户的过程中可能会（进行拷贝/移动并）抛出异常，而此时那个值已经从数据结构里被移除了，就导致那个元素一去不复返了。</li>
<li>但是 <code>.top()</code> 和 <code>.pop()</code> 分离的设计就自带 data race，用户想要取出顶部元素，于是就先调用 <code>.top()</code> 再调用 <code>.pop()</code>，然后托多线程的福两个同样的操作 interleaving 一下，就不对了。</li>
<li>解决方法是，依旧把 <code>.top()</code> 和 <code>.pop()</code> 合并为一个方法，同时想办法避免返回时异常导致数据丢失的问题：
<ul>
<li>方案1：加一个参数让用户传引用进来，这样的话“把值给用户”这个操作可以在从数据结构里移除元素之前完成，这样就不会有数据丢失的问题了。</li>
<li>方案2：提前把返回值用（拷贝/移动时不会抛异常的）  <code>std::shared_ptr</code> 装起来，然后返回给用户。</li>
<li>方案3：常规实现，但是要求用户定义的返回类型有无异常的拷贝或者构造函数（）。</li>
<li>综合方案：方案3肯定不能单独用，方案一最好要有（因为 overhead 最小），所以推荐的是方案 1+2 或者 1+3 组合。</li>
</ul>
</li>
</ul>
<p>避免死锁的一些建议：</p>
<ul>
<li>尽量不要嵌套锁。</li>
<li>需要同时锁多把锁时尽量一次锁完（<code>std::lock</code> 或 <code>std::scoped_lock</code>），不行的话就保证锁的顺序。</li>
<li>尽量不要在持有锁的时候调用户代码，因为用户可能也在获取锁。</li>
</ul>
<p><code>std::mutex</code> 的一些替代品：</p>
<ul>
<li>用于 lazy 初始化的 <code>std::once_flag</code> 和 <code>std::call_once</code>。</li>
<li>适合不怎么写的数据结构的读写锁 <code>boost:shared_mutex</code>。</li>
<li>支持递归的 <code>std::recursive_mutex</code>，不过在使用之前需要好好想想能不能避免递归锁。</li>
</ul>
<h2>Chapter 4</h2>
<p><code>std::future</code> 相关：</p>
<ul>
<li>
<code>std::async</code> 类似于 <code>std::thread</code> 但是通过返回 <code>std::future</code> 来提供返回值，可以通过参数来指定运行时机（另开线程或者在需要值的时候计算）。</li>
<li>
<code>std::promise</code> 属于纯手动，可以通过 <code>set_value()/set_value_at_thread_exit()</code> 设置值，也可以通过 <code>get_future()</code> 来获得 <code>std::future</code> 对象。</li>
<li>
<code>std::packaged_task</code> 可以把一个函数（可调用对象）和一个 <code>std::future</code> 相关联。执行（调用 <code>operator ()</code>）之后便可以从 <code>get_future()</code> 拿到的 <code>std::future</code> 对象里取出返回值。</li>
<li>
<code>std::future</code> 里可以存放异常，在使用 <code>get()</code> 时会把异常重新抛出。如果使用 <code>std::promise</code> 的话可以通过 <code>set_exception()/set_exception_at_thread_exit()</code> 来存放异常。</li>
<li>如果想多个线程都从 future 对象里取值的话要使用 <code>std::shared_future</code>，且每个线程都持有自己的 <code>std::shared_future</code> 的 copy。</li>
</ul>
<p><code>std::chrono</code> 转换行为：</p>
<ul>
<li>
<code>std::chrono::duration_cast</code> 的默认行为是截断而不是四舍五入。</li>
<li>
<code>std::duration</code> 可以自动从精度高的转成精度低的。</li>
</ul>
<p>支持 timeout 的好东西（其中 <code>xxx</code> 可以是 <code>for</code> 或者 <code>until</code>）：</p>
<ul>
<li><code>std::this_thread::sleep_xxx</code></li>
<li><code>std::condition_variable.wait_xxx</code></li>
<li><code>std::timed_mutex.try_lock_xxx</code></li>
<li><code>std::future.wait_xxx</code></li>
</ul>
<p><code>std::experimental</code> 里的好东西：</p>
<ul>
<li>
<code>std::experimental::future</code> 支持 continuation，就是那种老式的 JS 异步写法，靠 <code>.then()</code> 连在一起。</li>
<li>
<code>std::experimental::when_all/when_any()</code>，就是 JS 里的 <code>Promise.all/any()</code>。</li>
<li>
<code>std::experimental::latch</code> 就是 Go 里面的 <code>sync.WaitGroup</code>，感觉还是挺有用的。</li>
<li>
<code>std::experimental::barrier</code> 和 <code>std::experimental::latch</code> 在功能上有些类似，都可以用于等一些 threads 完成某件事
<ul>
<li>区别之一，所有 thread 到达 barrier 后，可以再次使用 barrier，而 latch 就保持 ready 状态。</li>
<li>区别之二，latch 可以等别的 thread 完成而自己不参与任务，但是 barrier 里执行任务和等待的是同一组 group。</li>
<li>还有一个 <code>std::experimental::flex_barrier</code> 支持构造时传入一个函数，用于到达后执行并修改等待的 thread 数量。</li>
</ul>
</li>
</ul>
<h2>Chapter 5</h2>
<p>注：这一章主要参考 cppreference 了，原因之一是 strongly happens-before 的概念在 C++ 20 中改变了。</p>
<p>一些和执行顺序相关的概念：</p>
<ul>
<li>注：先不考虑 <code>memory_order_consume</code>，因为 ISO C++ 委员会目前不建议使用这个。</li>
<li>sequenced-before 指的是同一个 thread 里根据 evaluation order 确定的执行顺序。</li>
<li>synchronizes-with 指的是对于一个 atomic，在 thread A 中的 store 是一个 release operation，而在 thread B 中的 load 是一个 acquire operation，那么这对 store 和 load 有 synchronizes-with 关系。</li>
<li>happens-before（在不考虑 consume 的情况下等同于 simply happens-before），简单来说它由一些 sequenced-before 和  synchronizes-before 连接而成。
<ul>
<li>strongly happens-before 在 happens-before 的基础上增加了对于 synchronizes-before 用到的 ordering 为 sequentially consistent 的要求，或者有夹在 sequenced-before 中间的任意 synchronizes-with。它能够帮助构建一个 total order（或者说那些东西是 linearizable）的。</li>
</ul>
</li>
</ul>
<p><code>std::memory_order</code> 的一些选项：</p>
<ul>
<li>
<code>memory_order_relaxed</code> 关于执行顺序什么也没保证。不过作为原子操作，还是能保证不会修改了一半被别人看到（atomicity），以及在看到后面修改的值之后不会再看到前面的值（consistent modification order）。</li>
<li>
<code>memory_order_acquire</code> 和 <code>memory_order_release</code>：
<ul>
<li>效果上，如果一个标记为 release 的 load 读到了来自一个被标记为 acquire 的 store 的值，那么 store 之前发生的（也就是 happens-before）的操作都可见了。</li>
<li>实现上不严谨地说，相当于不能把操作从 acquire store 的前面放到后面，也不能把操作从 release load 的后面放到前面。</li>
</ul>
</li>
<li>
<code>memory_order_seq_cst</code> 在 acquire 和 release 之上提供了一个全局的操作顺序。</li>
</ul>
<p><code>std::atomic</code> 系列：</p>
<ul>
<li>可以通过 <code>is_lock_free()</code> 来检查这个 atomic 是不是用锁模拟出来的。</li>
<li>
<code>std::atomic_flag</code> 是唯一一个确保没有 lock 的 atomic 类型，但是它支持的操作比较少。</li>
<li>
<code>compare_exchange_weak</code> 和 <code>compare_exchange_strong</code> 的区别在于 <code>compare_exchange_weak</code> 即使满足条件（atomic 中的值等于 expect）也不一定成功。通常如果在循环里使用就用 weak 版本，否则就 strong 版本。</li>
<li>想要原子性地使用 <code>std::shared_ptr</code> 可以用重载版本的 <code>std::atomic_load</code> 之类的函数，C++20 之后直接用 <code>std::atomic&lt;std::shared_ptr&lt;T&gt;&gt;</code>。</li>
</ul>
<h2>Chapter 6</h2>
<p>设计带锁数据结构时的一些指导性意见：</p>
<ul>
<li>保证数据结构的不变量被破坏时（或者说数据结构修改的中间状态）对于其它线程不可见。</li>
<li>设计 API 时注意避免潜在的 race condition，提供完整的操作而不是让用户去拼。（参考 Chapter 3 中的例子。）</li>
<li>在有异常发生的情况下也不要破坏不变量。（同样可以参考 Chapter 3 中的例子。）</li>
<li>通过减少锁的范围以及避免嵌套锁来降低死锁的概率。</li>
</ul>
<h2>Chapter 7</h2>
<p>lock-free 数据结构的特点：</p>
<ul>
<li>优势在于可以最大化并行程度，以及提升鲁棒性。对于有锁的数据结构，如果拿着锁的线程挂了可能会导致数据结构处于不一致的状态。</li>
<li>可能会出现 live lock 的问题，或者说有两个线程都因为对方而反复重启。</li>
<li>性能不一定好，比如因为更多数量的原子操作（相比起锁），或者缓存的反复失效。 </li>
<li>更难设计。</li>
</ul>
<p>lock-free 的一些难点（以书中的 stack 为例，代码就不贴了）：</p>
<ul>
<li>内存管理：
<ul>
<li>问题：<code>pop()</code> 中需要回收 <code>old_head</code>，但是问题是可能有别的 thread 也持有同一个 <code>old_head</code>，而且在 <code>head.compare_exchange_weak(old_head,old_head-&gt;next)</code> 的循环里，而这个循环访问了 <code>old_head</code>的 <code>next</code> 指针。所以说在 <code>pop()</code> 时释放 <code>old_head</code> 肯定是不对的。</li>
<li>方案1：先把需要删除的 node 存起来，等某个 <code>pop()</code> 发现没有别人也在 <code>pop()</code> 里的时候，把之前存下来的 node 一并释放。这么搞缺点也很明显，内存释放的时机很不可控，如果运气不好可能就永远都释放不掉。</li>
<li>方案2：拿个东西存一下所有 threads 正在访问的那个指针，然后离开 <code>pop()</code> 时根据有没有别的 thread 访问然后选择立刻释放或者延迟释放。</li>
<li>方案3：引用计数。如果 <code>std::atomic&lt;std::shared_ptr&lt;T&gt;&gt;</code> 是 lock-free 的，那么就还挺方便的，否则就手动实现。</li>
<li>注：可以参考一下 Linux Kernel 里的无锁同步机制 RCU (Read, Copy, Update)。</li>
</ul>
</li>
<li>ABA 问题：
<ul>
<li>简单来说，一次 <code>compare_exchange</code> 成功了确实能保证之前的值是你想要的，但并不保证是同一个东西。比如说你 exchange 了一个指针，然后那个指针被其它 thread 释放了然后又有新的 node 复用了那个地址。</li>
<li>一个解决方案是给那些指针附一个计数器，每次你复用地址时给计数器 +1 来保证值的不同。</li>
</ul>
</li>
</ul>
<h2>Chapter 8</h2>
<p>一些影响并行性能的点：</p>
<ul>
<li>核的数量（以及缓存的共享情况）。</li>
<li>Cache ping-pong（中文大概是缓存颠簸？）。如果多个线程经常需要同时修改一块内存区域，会导致经常出现因缓存失效导致的频繁缓存转移。</li>
<li>Cache false sharing（缓存伪共享）。表面上两个线程在操作不同的内存区域，但是因为靠得近所以在同一个缓存行里，便又导致了 Cache ping-pong。</li>
<li>内存访问的时空连续性，也就是对于某个 thread，它所访问的内存区域是分散还是集中。而在多线程应用中，减少所需要的缓存行数量可以降低 false sharing 概率和转移缓存行的代价。</li>
<li>线程数量和线程切换开销。如果没控制好数量导致处于就绪态的线程过多，就浪费资源在线程切换上了。</li>
</ul>
<h2>Chapter 10</h2>
<p>STL algorithm 里的 execution policy 参数：</p>
<ul>
<li>这个参数是 permission 而非 requirement。</li>
<li>出现异常的时候会 <code>std::terminate</code>，<code>std::bad_alloc</code> 除外。</li>
<li>渐进复杂度不变，但是常数会变大。</li>
<li>
<code>sequenced_policy</code> 是最普通的串行。<code>unsequenced_policy</code> 指可以向量化，<code>parallel_policy</code> 指可以在别的线程里搞，<code>parallel_unsequenced_policy</code> 在前面两者的基础上还支持在线程之间迁移。</li>
<li>其中 unsequenced 会要求用户代码 vectorization-safe，也就是说在同一个线程里同时执行的话不会有问题（反例就是拿锁，或者依赖前一个 iteration 的结果）。</li>
</ul>
<h2>Reference</h2>
<ul>
<li>C++ Concurrency in Action (2nd edition)</li>
<li><a rel="noopener" href="https://www.youtube.com/watch?v=UsSmA62eQwY" target="_blank">Concurrency in C++20 and beyond</a></li>
<li>https://ift.tt/tMjLGsQ</li>
</ul>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via Zerol's Notes https://zerol.me<br>
<br>
May 10, 2024 at 09:59AM
